### DISCUSSION TITLE
Type hints for unhandled exceptions code

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [python/typing#1476] [Q&A] [answered]
Below is a function to catch unhandled exceptions using excepthook. 
```
import sys
import logging

logger = logging.getLogger(__name__)

# A unhandled exception handler with excepthook
def handle_unhandled_exception(exc_type, exc_value, exc_traceback) -> None:
    if issubclass(exc_type, KeyboardInterrupt):
        # call the default excepthook
        sys.__excepthook__(exc_type, exc_value, exc_traceback)
        return
    # log message
    logger.critical('Unhandled exception:', exc_info=(exc_type, exc_value, exc_traceback))

# assign excepthook to the unhandled exception handler
sys.excepthook = handle_unhandled_exception
```
Running mypy generates the error:
```
> unhandled.py:7: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
```
What is the correct way to type the traceback parameters in the function? Also, is None the correct return type? Thank-you :)

**Comment by erictraut** [comment, marked as answer]
```python
from types import TracebackType

def handle_unhandled_exception(
    exc_type: type[BaseException],
    exc_value: BaseException,
    exc_traceback: TracebackType | None,
) -> None:
    ...
```

**Comment by hauntsaninja** [comment]
In addition to Eric's answer, two things that could help you figure this out yourself next time:
1) Check the official typeshed types: https://github.com/python/typeshed/blob/c5dde1e7208c85cddd6cc90e8c4af733d34d1cba/stdlib/sys.pyi#L35
2) Add `object` types to your arguments and see what mypy has to say: `def handle_unhandled_exception(exc_type: object, exc_value: object, exc_traceback: object) -> None:`
```
...
main.py:10: error: Argument 1 to "__excepthook__" has incompatible type "object"; expected "type[BaseException]"  [arg-type]
main.py:10: error: Argument 2 to "__excepthook__" has incompatible type "object"; expected "BaseException"  [arg-type]
main.py:10: error: Argument 3 to "__excepthook__" has incompatible type "object"; expected "TracebackType | None"  [arg-type]
...
```

**Reply by dineshbvadhia** [reply]
Ah, I'll definitely remember this. Wish I'd known it earlier as have just spent weeks typing my code :( 
Thank-you.

**Reply by Akuli** [reply]
Sometimes using `int` or some other obviously wrong type instead of `object` works better. For example, with `def handle_unhandled_exception(exc_type: int, exc_value: int, exc_traceback: int)` I get:

```
...
a.py:16: error: Incompatible types in assignment (expression has type "Callable[[int, int, int], None]", variable has type "Callable[[type[BaseException], BaseException, Optional[TracebackType]], Any]")
```

This error comes from the line that does `sys.excepthook = handle_unhandled_exception`, so the error message shows the types that the callback function needs to have. With `object` annotations, you actually see the types that `sys.__excepthook__` expects, which happen to be the same types.

You could also use `reveal_type(sys.excepthook)`.

**Reply by dineshbvadhia** [reply]
@Akuli Thank-you so much :)
